### PR TITLE
Change CPUCFSQuotaPeriod default value from 100ms to 100us to match Linux default

### PR DESCRIPTION
**Pull request by paskal**
cpu.cfs_period_us is 100μs by default despite having an "ms" unit for some unfortunate reason. Documentation: https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management

The desired effect of that change is to match k8s default `CPUCFSQuotaPeriod` value (100ms before that change) with one used in k8s when `CPUCFSQuotaPeriod` is not set, and Linux CFS (100μs, 1000x smaller than 100ms).

This PR is a followup of the #63437, which introduced the default of 100ms in k8s v1.12.

/kind api-change

#### Does this PR introduce a user-facing change? Yes, described above.

```release-note
[kubelet] Change default `cpuCFSQuotaPeriod` value with enabled `cpuCFSQuotaPeriod` flag from 100ms to 100µs to match the Linux CFS and k8s defaults. `cpuCFSQuotaPeriod` of 100ms now requires `customCPUCFSQuotaPeriod` flag to be set to work.
```

### CODE REVIEW

File: pkg/kubelet/apis/config/validation/validation.go
Comment (szuecs) [inline pkg/kubelet/apis/config/validation/validation.go:74]: I think `usec` instead of `µs` is fine here for consistency

File: pkg/kubelet/apis/config/validation/validation_test.go
Comment (szuecs) [inline pkg/kubelet/apis/config/validation/validation_test.go:159]: I think usec instead of µs is fine here for consistency

File: pkg/kubelet/cm/helpers_linux.go
Comment (szuecs) [inline pkg/kubelet/cm/helpers_linux.go:56]: I think this should change the first part of the comment and not add more text.

File: pkg/kubelet/cm/pod_container_manager_linux.go
Comment (szuecs) [inline pkg/kubelet/cm/pod_container_manager_linux.go:56]: I think usec instead of µs is fine here for consistency

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (szuecs) [inline pkg/kubelet/kuberuntime/helpers_linux.go:25]: I think usec instead of µs is fine here for consistency

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (szuecs) [inline pkg/kubelet/kuberuntime/helpers_linux.go:34]: I think this should change the first part of the comment and not add more text.
and
I think usec instead of µs is fine here for consistency

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (szuecs) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:117]: I think usec instead of µs is fine here for consistency

File: cmd/kubelet/app/options/options.go
Comment (liggitt) [inline cmd/kubelet/app/options/options.go:468]: this makes a super confusing message:

> Sets CPU CFS quota period value, cpu.cfs_period_us, defaults to Linux Kernel default (100μs) (default 100ms)

Comment (liggitt) [reply cmd/kubelet/app/options/options.go:468]: I'm also not sure this change is correct... this is a time.Duration, and is currently defaulted to 100 milliseconds:

```go
	if obj.CPUCFSQuotaPeriod == nil {
		obj.CPUCFSQuotaPeriod = &metav1.Duration{Duration: 100 * time.Millisecond}
	}
```

Comment (paskal) [reply cmd/kubelet/app/options/options.go]: It is supposed to be 100 microseconds to be in sync with Linux Kernel. Thanks a lot for noticing it! I've changed the line you mentioned.

As far as I understand, you need to enable the flag `CustomCPUCFSQuotaPeriod` to use that value, and if you set a non-default `cpuCFSQuotaPeriod`, nothing changes for you.
However, if you have `CustomCPUCFSQuotaPeriod` enabled but haven't set `cpuCFSQuotaPeriod`, after my change, it will become in sync with Linux Kernel default 1000x smaller, 100µs instead of 100ms. That is desirable behaviour, but it's magnitudes different from just the documentation change I was thinking about initially.

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (liggitt) [inline pkg/kubelet/apis/config/v1beta1/defaults.go:191]: Changing defaults is generally not allowed and is considered an incompatible change.

It looks like the current 100ms default has been in place since 1.12 (https://github.com/kubernetes/kubernetes/pull/63437)

I don't have context to know what the implications of making the default 1000x shorter are on all existing clusters. cc @kubernetes/sig-node-api-reviews

Comment (paskal) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: To clarify, by default clusters still use [the 100us value](https://github.com/kubernetes/kubernetes/blob/37761a3/pkg/kubelet/cm/helpers_linux.go#L48-L67) which is [in sync with Linux](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management).

Enabling `CustomCPUCFSQuotaPeriod` flag alongside providing a custom `cpuCFSQuotaPeriod` value leaves the cluster owner with the selected value.

Only clusters with `CustomCPUCFSQuotaPeriod` enabled but without `cpuCFSQuotaPeriod` would be affected and receive 1000x shorter value.

I'm as upset as you are about this change, but I think it has to be done. The improper default value for `cpuCFSQuotaPeriod` introduces even more confusion to an already complicated topic (see #51135 for context).

File: cmd/kubelet/app/options/options.go
Comment (szuecs) [reply cmd/kubelet/app/options/options.go]: @liggitt I think in my pr I assumed that all the code comments were correct and these were 100ms (we even had the production problem with not being able to have lower latency than 100ms), so I set the default to 100ms. However it's only used if the featureGate is enabled, and no value is set. I think it's rarely a combination setting that anyone would do.
Of course we can't be 100% sure, but tbh maybe it's a bug fix, that no one before observed.
@derekwaynecarr and @dims wdyt from sig-node point of view ?

Comment (liggitt) [reply cmd/kubelet/app/options/options.go]: > However it's only used if the featureGate is enabled, and no value is set. I think it's rarely a combination setting that anyone would do.

are all uses of the kubelet config value gated by the feature gate? if so, the fact that that is still an alpha feature gate makes it more acceptable to change this default, imo

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (bobbypage) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: It does look that if the `CustomCPUCFSQuotaPeriod` (still alpha) feature gate is not enabled, then the period will be read from the constant QuotaPeriod and ignore the kubelet config value for `CPUCFSQuotaPeriod` - xref: https://github.com/kubernetes/kubernetes/blob/f3d90aef8d4a0a6934a4a51705c241a86f66b00c/pkg/kubelet/cm/helpers_linux.go#L65-L67

It sounds like the problem is the original default here was invalid in the first place -- it should have been the same value as the linux kernel. And the original default confused `ms` to mean millisecond instead of microseconds :). The value should be microseconds, quoting from kernel docs - "cpu.cfs_quota_us: run-time replenished within a period (in microseconds)"

Additionally, we should consider if validation should be added for the kubelet config, enabling the `CustomCPUCFSQuotaPeriod` feature gate without specifying a `CPUCFSQuotaPeriod` should be invalid...

File: cmd/kubelet/app/options/options.go
Comment (szuecs) [reply cmd/kubelet/app/options/options.go]: yes all uses are gated by the feature gate

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (szuecs) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: @bobbypage yes that would be a great solution, then even if people have this wrong configuration they would get a notification before going silently in production with this.

Comment (paskal) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: I've altered this PR to include only changing the default value, and I updated the code and PR description (so please re-review).

Harmless documentation update, which just reflects the real state of code more precisely, was moved to #111554.

I might have an opinion about restricting users from enabling `CustomCPUCFSQuotaPeriod` but not setting `CPUCFSQuotaPeriod`, but I don't have arguments to support it. Tell me what to do, and I'll alter the PR as you say. If the vote is pro the change, please point me to any option which works in a similar way as I'm not familiar with that project code and the only way for me to do it right is to copy same code from another place.

Comment (bobbypage) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: hmm, thinking about it some more, the issue with adding the validation is if any users will currently enable the feature gate and also don't set a `CPUCFSQuotaPeriod` then their kubelet will not start up (as this would be a breaking change). Additionally, if the feature gate would ever be promoted to beta and it would be turned on by default we would not want the validation to fail if the a CPUCFSQuotaPeriod is not set. So thinking about it some more, I think we should just focus on making the default correct.

**Comment by bobbypage** [comment]
Should we update comment in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/helpers_linux.go#L47-L48 ?

**Comment by paskal** [comment]
> Should we update comment in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/helpers_linux.go#L47-L48 and comments in MilliCPUToQuota to rename s/ms/us?

Sure, @bobbypage, I've moved that change #111554 because it doesn't change anything but just clarifies the existing state of things, is harmless to the users, and should be easier to approve and merge. I would be very grateful if someone would help me get the necessary approvals for it, as I'm fairly confused by the flow.

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (szuecs) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: What if we first make it fail and 3 versions later make the default correct?

Comment (liggitt) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: Code should be written in a way that it will be correct when the feature gate graduates to GA, is unconditionally enabled, and removed in the future. Requiring a custom value if the feature gate is enabled doesn't match that.

Changing the default while the feature is in the alpha state seems reasonable and sufficient to me

Comment (szuecs) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:191]: Thanks @liggitt

**Comment by liggitt** [comment]
API change lgtm
/hold for node approval

**Comment by paskal** [comment]
I have no idea why the e2e tests are failing with `kubelet` not being ready in time, I would greatly appreciate any help. I don't know what part of the change could cause that problem.

**Comment by bobbypage** [comment]
It looks like the unit tests are still failing - https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/111520/pull-kubernetes-unit/1554556782813122560

e.g. 

```
=== Failed
=== FAIL: pkg/kubelet/apis/config/scheme TestRoundTripYAML/KubeletConfiguration_v1beta1Tov1beta1_default (0.02s)
    helpers.go:90: Output does not match expected, diff (- want, + got):
          strings.Join({
          	... // 370 identical bytes
          	"gMaxFiles: 5\ncontainerLogMaxSize: 10Mi\ncontentType: application/",
          	"vnd.kubernetes.protobuf\ncpuCFSQuota: true\ncpuCFSQuotaPeriod: 100",
        - 	"u",
        + 	"µ",
          	"s\ncpuManagerPolicy: none\ncpuManagerReconcilePeriod: 10s\nenableCo",
          	"ntrollerAttachDetach: true\nenableDebugFlagsHandler: true\nenableD",
          	... // 1501 identical bytes
          }, "")
        
    helpers.go:105: if the diff is expected because of a new type or a new field, re-run with UPDATE_COMPONENTCONFIG_FIXTURE_DATA=true to update the compatibility data or generate missing files
    --- FAIL: TestRoundTripYAML/KubeletConfiguration_v1beta1Tov1beta1_default (0.02s)
```

**Comment by paskal** [comment]
Unit tests failed due to yaml file having `us` instead of `µs`, I fixed it now. I don't know why e2e tests are failing and how to fix them. Reverting yaml files to `100ms` produce same failure in e2e tests, so they are not relevant.

**Comment by bobbypage** [comment]
> Unit tests failed due to yaml file having `us` instead of `µs`, I fixed it now. I don't know why e2e tests are failing and how to fix them. Reverting yaml files to `100ms` produce same failure in e2e tests, so they are not relevant.

E2E tests need to be passing to get the PR merged.

Please see here - https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/111520/pull-kubernetes-e2e-gce-ubuntu-containerd/1554830092503355392/artifacts/e2e-e2a60cb2ac-a7d53-minion-group-05r1/kubelet.log

Kubelet is failing to start because:

```
Aug 03 14:35:19.780469 e2e-e2a60cb2ac-a7d53-minion-group-05r1 kubelet[4188]: E0803 14:35:19.772966    4188 run.go:74] "command failed" err="failed to validate kubelet configuration, error: invalid configuration: cpuCFSQuotaPeriod (--cpu-cfs-quota-period) {100µs} requires feature gate CustomCPUCFSQuotaPeriod, path: &TypeMeta{Kind:,APIVersion:,}"
```

**Comment by paskal** [comment]
Thanks a lot for pointer to the logs location! I'll fix it, but I'm struggling to understand of why my changes triggered that problem, as I only changed existing value and haven't introduced configuration overrides anywhere in the code.

**Comment by paskal** [comment]
I'm clueless about how to properly fix the error which is highlighted by CI now: for some reason thousands of lines supposed to be added to `api/openapi-spec/swagger.json` and others like `api/openapi-spec/v3/apis__autoscaling__v2beta2_openapi.json`. For me as a naive contributor it looks like I've triggered generation for whole new version of APIs. Can someone please tell me what is the right course of actions here?

**Comment by liggitt** [comment]
It was an unrelated failure now resolved in master

/retest

**Comment by paskal** [comment]
Thanks!

Reviewers, please review the code and the changelog entry part in the PR description: the PR changed the scope since the original introduction and requires a proper review before merging. https://github.com/kubernetes/kubernetes/pull/111554 (code-comments-only change) should be reviewed as well.

**Review by szuecs — APPROVED** [review]

**Comment by paskal** [comment]
@bobbypage, could you please tell me if something is still missing here and in #111554, which prevents merging them into master? I'm not familiar with k8s merge flow enough to know what's missing at this moment here, and in another PR, approval is missing, but I already pinged all approvers once, and nothing happened.

**Comment by PixelOrange** [comment]
@paskal It looks to me like there are three people still pending approval on this one:

bobbypage
liggitt
mrunalp

It also says tide is pending and isn't mergeable, although I am not familiar with tide. I hope they see it soon, I've been anxiously awaiting this change since you first opened it.

**Review by endocrimes — COMMENTED** [review]
/lgtm

it'll merge when we thaw the codebase after the release of 1.25.

**Comment by paskal** [comment]
For history, I was wrong, and the default value is 100ms and not 100μs: PR #112077 reverted these breaking changes, and PR #112123 clarified the code to prevent mistakes about units of measurement around this feature in the future. I'll try clarifying [linked](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management) kernel.org documentation as well.

**Comment by PixelOrange** [comment]
@paskal I read through your discussion in #112108 and while I now also follow, I believe that the setting should not include a unit if the default value is a conflicting unit. If the value can never be less than 1ms then there's never a reason for microseconds to come into play. I think the "_us" should be removed from the setting and the current description stating the range of valid values be left unchanged.

Unfortunately, I don't know how to go about requesting that change. I'm coming at this from the same perspective as many other people who've stumbled on these conversations due to poorly performing applications under CFS. What do you suggest?

**Comment by paskal** [comment]
@PixelOrange, I would propose you check changes in #112077 and verify if something is missing there from your perspective. The valid range of values is 1ms-1s, and the k8s source code will be updated to reflect it.

`_us` postfix comes into play only in the Linux kernel and is part of the public API, meaning that logic will never change. The only thing which could be changed is documentation around the topic, and I'll try to come up with that change.

**Comment by PixelOrange** [comment]
Everything looks good as far as I can tell.

That's unfortunate about the public API. I found an article on how to submit kernel patches but I imagine that's a daunting task.

If you want a second set of eyes for the documentation updates, please feel free to message me.

**Comment by szuecs** [comment]
Thanks @paskal for going the long way.

**Pull request merged**
